File: Current/python_code/BlackScholes.py
import numpy as np

class BlackScholes:
    def __init__(self, interest_, volatility_):
        self.interest = interest_
        self.volatility = volatility_


class Stock(BlackScholes):
    def __init__(self, interest_, volatility_, initialPrice_):
        self.interest = interest_
        self.volatility = volatility_
        self.initialPrice = initialPrice_
    
    def generateSamplePath(self, expiry, steps):
        """Generates a sample path"""
        stockPrice = self.initialPrice
        samplePath = np.zeros(steps)
        timeIncrement = expiry/steps
        samplePath[0] = stockPrice
        for i in range(1, steps):
            noise = np.random.normal(0, 1)
            samplePath[i] = stockPrice*np.exp((self.interest - 0.5*self.volatility**2)*timeIncrement + self.volatility*np.sqrt(timeIncrement)*noise)
            stockPrice = samplePath[i]
        self.stockProcess = samplePath
    
    def getSamplePath(self):
        return self.stockProcess

File: Current/cpp_code/BlackScholes.cpp
#include <cmath>
#include "BlackScholes.h"


const double pi = 4.0*atan(1.0);

double gaussian(){
    // Generates a Gaussian Random Variable with mean 0, variance 1
    double U1 = (rand()+1.0)/(RAND_MAX+1.0);
    double U2 = (rand()+1.0)/(RAND_MAX+1.0);
    return sqrt(-2.0*log(U1))*cos(2.0*pi*U2);
}

BlackScholes::BlackScholes(
        double initialPrice_,
        double interest_,
        double volatility_
        ){
        initialPrice = initialPrice_;
        interest = interest_;
        volatility = volatility_;
        srand(time(NULL));
        // sets the seed for random numbers
    }

void BlackScholes::generateSamplePath(double T, int m, SamplePath& S){
    double stockPrice = initialPrice;
    for (int k=0; k<m; k++){
        S[k] = stockPrice*exp((interest-volatility*volatility*0.5)*(T/m)+volatility*sqrt(T/m)*gaussian());
        stockPrice = S[k];
    }
}

// int main(){
//     return 0;
// }
